include/helib/query.h

```diff
@@ -515,6 +603,126 @@ class QueryBuilder
+    vecvec Fs(expr.size());
+    {
+      std::vector<long> v(columns);
+      std::iota(v.begin(), v.end(), 0);
+      std::fill(Fs.begin(), Fs.end(), v);
+    }
+    std::vector<long> mus(expr.size(), 1);
+    std::vector<Matrix<long>> taus;
+    taus.reserve(expr.size());
+
+    // Create the taus
+    for (size_t i = 0; i < expr.size(); ++i) { // Each tau
+      assertTrue(expr[i].size() > 0,
+                 "empty clause found at index " + std::to_string(i) +
+                     ", function called without Tidy().");
+      mus[i] = 0;                 // Set mu to zero.
+      Matrix<long> M(columns, 1); // Create temp tau matrix
+      containsOR = (expr[i].size() > 1) ? true : containsOR;
+      for (size_t j = 0; j < expr[i].size(); ++j) // Each column index
+      {
+        // column should not already have been seen in this clause
+        assertTrue(M(std::abs(expr[i][j]) - 1, 0) == 0,
+                   "Column " + std::to_string(std::abs(expr[i][j]) - 1) +
+                       " repeated in clause " + std::to_string(i) +
+                       ", function called without Tidy().");
+        // add one to the offset for each NOT
+        if (expr[i][j] < 0)
+          mus[i] += 1;
+        // mark NOT columns as -1, columns as 1
+        M(std::abs(expr[i][j]) - 1, 0) = ((expr[i][j] >= 0) ? 1 : -1);
+      }
+      taus.push_back(std::move(M));
+    }
+    return QueryType(std::move(Fs),
+                     std::move(mus),
+                     std::move(taus),
+                     containsOR);
+  }
+  /**
+   * @brief Given a vecvec intepreted as an AND of ORs, return a vecvec
+   * corresponding to the negation. Called by expandOr()
+   */
+  vecvec negate(const vecvec& clause) const
+  {
+    // We build the negation clause by clause
+    vecvec notclause = {{}};
+    for (const auto& i : clause) {
+      // i = {a,b,c,...}. Negate this, to give {{-a},{-b},{-c},...}
+      vecvec nextclause;
+      for (const auto& j : i) {
+        nextclause.push_back({-1 * j});
+      }
+      vecvec notclausetemp;
+      notclausetemp.reserve(nextclause.size() * notclause.size());
+      // then Cartesian style product
+      for (const auto& j : notclause) {
+        for (const auto& k : nextclause) {
+          std::vector<long> x = j;
+          x.insert(x.end(), k.begin(), k.end());
+          notclausetemp.push_back(x);
+        }
+      }
+      notclause = notclausetemp;
+    }
+    return notclause;
+  }
+  /*
+  tidies an inner clause in place by deleting duplicate columns and removing
+  columns for which both the column and it's negation appear. Called by Tidy()
+  */
+  void tidyClause(std::vector<long>& clause) const
+  {
+    std::unordered_set<long> vars;
+    std::vector<long> newclause;
+    for (auto& i : clause) {
+      if (vars.find(i) == vars.end()) {
+        newclause.push_back(i);
+        vars.insert(i);
+      } else {
+        continue;
+      }
+    }
+    clause.clear();
+    for (auto& i : newclause) {
```

Comment:
```suggestion
    for (const auto& i : newclause) {
```

Reply:
I'm not 100% sure about this one. Need to check if it effects line 722 `vars.find(-1 * i)`.

Reply:
works fine, thanks!